src/streamUtils/BridgeDuplex.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright 2019 Centrality Investments Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import v4 from 'uuid/v4';
+import { browser } from 'webextension-polyfill-ts';
+import {
+  MessageOrigin,
+} from '../types';
+import {InPageMsgTypes} from '../types/message';
+import { MultiplexWindowMessageDuplex } from './MultiplexWindowMessageDuplex';
+import { RuntimePortDuplex } from './RuntimePortDuplex';
+
+/**
+ * BridgeDuplex is a bridge duplex stream using in content script for connecting background and inject
+ */
+export class BridgeDuplex extends MultiplexWindowMessageDuplex {
+  portStream: RuntimePortDuplex;
+
+  constructor(protected window: Window, tag: string, untag?: string) {
+    super(window, tag, untag);
+  }
+
+  protected eventHandler = ({ data }: MessageEvent) => {
+    if (!this.portStream && data.type ===InPageMsgTypes.ENABLE) {
+      const port = browser.runtime.connect(null, {
```

Comment:
make port a singleton

Reply:
If port is singleton, all contents will share one port, while means their port name will be the same. We want to use port name as the identity to filter contents. So maybe we should not make port a singleton